### PR TITLE
Reorgonize Helmholtz EOS and Add Span-Wagner CO2

### PR DESCRIPTION
**Pull request by eslickj**
This adds Span-Wangner and reorganizes the IAPWS-95 stuff into a fairly general Helmholtz EOS, so it should be easy to add additional substances now.

### CODE REVIEW

File: src/helmholtz/common/helmholtz_external.cpp [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+      hes[0] = 0;
+      hes[1] = 0;
+      hes[2] = 0;}}
+  return w(delta, tau);
+}
+
+s_real sat_delta_liq_with_derivs(s_real tau, s_real *grad, s_real *hes){
+  // For efficency I'm requiring memoization here.  Delta_v and delta_l are
+  // calculated together so should just remember both.  This will break if you
+  // turn off memoiztion.
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_external.cpp]: Typo: memoization

File: src/helmholtz/common/helmholtz_external.cpp [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+  // Should be there now
+  val = memoize::get_un(memoize::DL_SAT_FUNC, tau, grad, hes);
+  if(!std::isnan(val)) return val;
+  return (s_real)NAN;
+}
+
+s_real sat_delta_vap_with_derivs(s_real tau, s_real *grad, s_real *hes){
+  // For efficency I'm requiring memoization here.  Delta_v and delta_l are
+  // calculated together so should just remember both.  This will break if you
+  // turn off memoiztion.
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_external.cpp]: Typo: efficiency and memoization

File: src/helmholtz/common/helmholtz_external.cpp [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+  // Should be there now
+  val = memoize::get_un(memoize::DV_SAT_FUNC, tau, grad, hes);
+  if(!std::isnan(val)) return val;
+  return (s_real)NAN;
+}
+
+s_real sat_p_with_derivs(s_real tau, s_real *grad, s_real *hes, bool limit){
+  //Before getting into the real calculation, check if outside the allowed range
+  //of 240K to T_c, the low end of this range doen't mean anything.  Its too cold
+  //to expect liquid, but the calucations hold up there so for numerical reasons
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_external.cpp]: Typo: calculations

File: src/helmholtz/common/helmholtz_external.cpp [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+  //I'll allow it.  Above the critical temperture there is only a single phase
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_external.cpp]: Typo: temperature

File: src/helmholtz/common/helmholtz_external.cpp [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+  if(free_grad) delete[] grad; // free grad and hes if not allocated by calling
+  if(free_hes) delete[] hes;   // function
+  return Psat;
+}
+
+s_real sat_tau_with_derivs(s_real pr, s_real *grad, s_real *hes, int *nit){
+  //Before getting into the real calculation, check if outside the allowed range
+  //of 240K to T_c, the low end of this range doen't mean anything.  Its too cold
+  //to expect liquid, but the calucations hold up there so for numerical reasons
+  //I'll allow it.  Above the critical temperture there is only a single phase
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_external.cpp]: Typo: calculations and temperature

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+------------------------------------------------------------------------------*/
+
+/*------------------------------------------------------------------------------
+ This file contains functions to solve for liquid and vapor denstiy from
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: density

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+#include<iostream>
+
+#include"helmholtz_memo.h"
+#include"helmholtz_config.h"
+#include"helmholtz_deriv_parts.h"
+#include"helmholtz_solve.h"
+
+/*------------------------------------------------------------------------------
+In this section you will find function for calculating density from pressure
+and teperature. This lets you calculate properties in the more standard way as
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: temperature

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+a function of temperature and pressure instead of density and pressure.
+
+Unfortunatly this is difficult and a bit messy.
+*-----------------------------------------------------------------------------*/
+s_real delta_p_tau_rf(s_real pr, s_real tau, s_real a, s_real b, bool bisect){
+  /*----------------------------------------------------------------------------
+  Bracketing methods, false position and bisection, for finding a better initial
+  guess for density when solving for density from temperature and pressure. This
+  is only used in particularly diffucult areas.  At this point it probably
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: difficult

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+            bisection isn't really used, but I kept it around for debugging
+    a: first density bound (kg/m3)  | really density, why didn't use delta?
+    b: second density bound (kg/m3) | it was easier to think in terms of density
+  Returns:
+    delta: reduced density at pr and tau (more or less approximate)
+  ----------------------------------------------------------------------------*/
+  s_real c=a, fa, fb, fc;
+  int it = 0;
+  // If right by critical point guess critical density. (okay this isn't part
+  // of a backeting method but it was conveneint).
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: bracketing? convenient

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+  if( fabs(T_c/tau - T_c) < 1e-7 && fabs(pr - P_c) < 1e-4) return 1;
+  // solve f(delta, tau) = 0; f(delta, tau) = p(delta, tau) - pr
+  a /= rho_c; // convert to reduced density
+  b /= rho_c; // convert to reduced density
+  fa = p(a, tau) - pr; // initial f(a, tau)
+  fb = p(b, tau) - pr; // initial f(b, tau)
+  while(it < MAX_IT_BRACKET && (a - b)*(a - b) > TOL_BRACKET){
+    if(bisect) c = (a + b)/2.0; // bisection
+    else c = b - fb*(b - a)/(fb - fa); //regula falsi
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: "regula falsi"?

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    fc = p(c, tau) - pr; // calcualte f(c)
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: calculate

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+  Args:
+   pr: pressure (kPa)
+   tau: inverse of reduced pressure Tc/T
+   delta0: initial guess for delta
+   tol: absolute residual tolerance for convergence
+   nit: pointer to return number of iterations to, or NULL
+   grad: location to return gradient of delta wrt pr and tau or NULL
+   hes: location to return hessian (upper triangle) or NULL
+  Returns:
+   delta: reduced density (accuracy depends on tolarance and function shape)
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: tolerance

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+  while(fabs(fun) > tol && it < MAX_IT_DELTA){
+    delta = delta - fun*gradp[0]/(gradp[0]*gradp[0] - 0.5*fun*hesp[0]);
+    fun = p_with_derivs(delta, tau, gradp, hesp) - pr;
+    ++it;
+  }
+  if(nit != NULL) *nit = it;
+  if(grad != NULL){ // calculate gradient if needed
+    grad[0] = 1.0/gradp[0];
+    grad[1] = -gradp[1]*grad[0]; //triple product
+    if(hes != NULL){ // calculate hession if needed.
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: Hessian

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+      hes[1] = -(hesp[1] + hesp[0]*grad[1])/gradp[0]/gradp[0];
+      hes[2] = -(grad[0]*(hesp[2] + grad[1]*hesp[1]) + gradp[1]*hes[1]);}}
+  return delta;
+}
+
+s_real delta_liq(s_real p, s_real tau, s_real *grad, s_real *hes, int *nit){
+  /*----------------------------------------------------------------------------
+  Get a good liquid or super critical inital density guess then call
+  delta_p_tau() to calculate density. In difficult cases an interval with the
+  desired soultion is used with a backeting method to produce a better inital
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: bracketing?
Typo: initial

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+  exact same inputs when using the Pyomo wrapper, this function is memoized.
+
+  Args:
+   p: pressure (kPa)
+   tau: inverse of reduced pressure Tc/T
+   grad: location to return gradient of delta wrt pr and tau or NULL
+   hes: location to return hessian (upper triangle) of NULL
+   nit: pointer to return number of iterations to, or NULL
+  Returns:
+   delta: reduced density (accuracy depends on tolarance and function shape)
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: tolerance

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+  bool free_grad = 0, free_hes = 0; // grad and/or hes not provided so allocate
+  // Since I'm going to cache results, grad and hes will get calculated
+  // whether requested or not.  If they were NULL allocate space.
+  if(grad==NULL){grad = new s_real[2]; free_grad = 1;}
+  if(hes==NULL){hes = new s_real[3]; free_hes = 1;}
+  s_real delta, delta0, tol=TOL_DELTA_LIQ;
+  delta0 = LIQUID_DELTA_GUESS;
+  delta = delta_p_tau(p, tau, delta0, tol, nit, grad, hes); //solve
+  if(std::isnan(delta) || delta < 1e-12 || delta > 5.0){
+    // This is just to avoid evaluation errors.  Want to be able to calucalte
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: calculate

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+    // vapor properties even when vapor doesn't exist.  In the IDAES Framework
+    // these properties may be calculated and multipled by a zero liquid fraction,
+    // so it doesn't mater that they are wrong.
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: matter

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+  }
+  memoize::add_bin(memoize::delta_liq, p, tau, delta, grad, hes); //store
+  if(free_grad) delete[] grad; // free grad and hes if not allocated by calling
+  if(free_hes) delete[] hes;   //   function
+  return delta;
+}
+
+s_real delta_vap(s_real p, s_real tau, s_real *grad, s_real *hes, int *nit){
+  /*----------------------------------------------------------------------------
+  Get a good vapor or super critical inital density guess then call
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: initial

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+  delta_p_tau() to calculate density. In the supercritical region this just
+  calls the liquid function. In the rest of the vapor region the inital guess
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: initial

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+  exact same inputs when using the Pyomo wrapper, this function is memoized.
+
+  Args:
+   p: pressure (kPa)
+   tau: inverse of reduced pressure Tc/T
+   grad: location to return gradient of delta wrt pr and tau or NULL
+   hes: location to return hessian (upper triangle) of NULL
+   nit: pointer to return number of iterations to, or NULL
+  Returns:
+   delta: reduced density (accuracy depends on tolarance and function shape)
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: tolerance

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+  // If supercritical use the liquid calculation, which includes sc region
+  //if(tau <= 1.0 || p >= P_c) return delta_liq(p, tau, grad, hes);
+  // Since I'm going to cache results, grad and hes will get calculated
+  // whether requested or not.  If they were NULL allocate space.
+  if(grad==NULL){grad = new s_real[2]; free_grad = 1;}
+  if(hes==NULL){hes = new s_real[3]; free_hes = 1;}
+  delta0 = VAPOR_DELTA_GUESS;
+  delta = delta_p_tau(p, tau, delta0, TOL_DELTA_VAP, nit, grad, hes);
+  if(std::isnan(delta) || delta < 1e-12 || delta > 5.0){
+    // This is just to avoid evaluation errors.  Want to be able to calucalte
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: calculate

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+    // vapor properties even when vapor doesn't exist.  In the IDAES Framework
+    // these properties may be calculated and multipled by a zero vapor fraction,
+    // so it doesn't mater that they are wrong.
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: matter

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+  memoize::add_bin(memoize::DV_FUNC, p, tau, delta, grad, hes); // store result
+  if(free_grad) delete[] grad; // free grad and hes if not allocated by calling
+  if(free_hes) delete[] hes; //   function
+  return delta;
+}
+
+
+/*------------------------------------------------------------------------------
+In this section you will find functions for calculating the saturation curve.
+Staturation pressure and density as a function of temperature.
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: saturation

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+*-----------------------------------------------------------------------------*/
+s_real sat_delta_liq(s_real tau){ //caculate saturated liquid density from tau
+  s_real delta_l, delta_v;
+  sat(tau, &delta_l, &delta_v);
+  return delta_l;
+}
+
+s_real sat_delta_vap(s_real tau){ //caculate saturated vapor density from tau
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: calculate

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+  return 2.0*phir_delta(delta, tau) + delta*phir_delta2(delta, tau) + 1.0/delta;
+}
+
+inline s_real Delta_Aka(s_real delta_l, s_real delta_v, s_real tau){
+  return J_delta(delta_v, tau)*K_delta(delta_l, tau) -
+         J_delta(delta_l, tau)*K_delta(delta_v, tau);
+}
+
+int sat(s_real tau, s_real *delta_l_sol, s_real *delta_v_sol){
+    //Get stautated phase densities at tau by Akasaka (2008) method
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: saturated

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+      delta_v = 1.0;
+      max_it=0;
+    }
+    else{
+      // okay so you've decided to solve this thing
+      delta_l = DELTA_LIQ_SAT_GUESS;
+      delta_v = DELTA_VAP_SAT_GUESS;
+    }
+    // Since the equilibrium conditions are gl = gv and pl = pv, I am using the
+    // the relative differnce in g as a convergence criteria, that is easy to
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: difference

File: src/helmholtz/common/helmholtz_solve.cpp [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+    // understand.  fg < tol for convergence, fg is calucalted upfront in the
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.cpp]: Typo: calculated

File: src/helmholtz/common/helmholtz_solve.h [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+#define _INCLUDE_HELMHOLTZ_SOLVE_H_
+
+#include "helmholtz_config.h"
+
+
+/*------------------------------------------------------------------------------
+  Function to calculate delta (rho/rho_c) from P and tau (T_c/T)
+------------------------------------------------------------------------------*/
+
+// Solve for delta using Halley's method delta0 in an inital guess, tol is
```
Comment (blnicho) [inline src/helmholtz/common/helmholtz_solve.h]: Typo: initial

File: src/helmholtz/iapws95/iapws95_guess.cpp [hunk 29]
```diff
@@ -0,0 +1,194 @@
+#include"helmholtz_solve.h"
+#include"iapws95_param.h"
+
+s_real p_sat_iapws97(s_real tau){ //saturation pressure from tau IAPWS-97 eq.
+  //the IAPWS-97 isn't as consitent as IAPWS-95, but this provides a good guess
```
Comment (blnicho) [inline src/helmholtz/iapws95/iapws95_guess.cpp]: Typo: consistent

File: src/helmholtz/iapws95/iapws95_guess.cpp [hunk 30]
```diff
@@ -0,0 +1,194 @@
...
+           - 1.75493479*pow(XX,16.0/3.0)
+           - 45.5170352*pow(XX,43.0/3.0)
+           - 6.74694450e5*pow(XX,110.0/3.0);
+  return delta;
+}
+
+s_real delta_p_tau_supercritical_guess_iapws95(s_real p, s_real tau){
+
+  /* This uses approximate isochors to get a reasonable initial guess for the
+  density solvution for a vapor at a particular temperature and pressure.  If
```
Comment (blnicho) [inline src/helmholtz/iapws95/iapws95_guess.cpp]: Typo: solution

File: src/helmholtz/iapws95/iapws95_guess.cpp [hunk 31]
```diff
@@ -0,0 +1,194 @@
...
+  else if(p < -417462.7784 + 1107.467347*T + 0.884115157*T*T){ //rho < 1000
+    return delta_p_tau_rf(p, tau, 895, 1005, 0);
+  }
+  return 1100/rho_c;
+}
+
+s_real delta_p_tau_vap_guess_iapws95(s_real p, s_real tau){
+
+  /* This uses approximate isochors to get a reasonable initial guess for the
+  density solvution for a vapor at a particular temperature and pressure.  If
```
Comment (blnicho) [inline src/helmholtz/iapws95/iapws95_guess.cpp]: Typo: solution

File: src/helmholtz/iapws95/iapws95_guess.cpp [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+  }
+  else if(p < -167178.3915 + 313.7993977*T - 0.033828951*T*T){ //rho < 300
+    return 190/rho_c;
+  }
+  return delta_p_tau_rf(p, tau, 295, rho_c, 0);
+}
+
+s_real delta_p_tau_liq_guess_iapws95(s_real p, s_real tau){
+  /* This uses approximate isochors to get a reasonable initial guess for the
+  density solvution for a vapor at a particular temperature and pressure.  If
```
Comment (blnicho) [inline src/helmholtz/iapws95/iapws95_guess.cpp]: Typo: solution

File: src/helmholtz/iapws95/iapws95_param.h [hunk 33]
```diff
@@ -36,14 +36,20 @@
...
 
 // Critiacal point for water and R
-static const s_real T_c = 647.096;   // Critical T (K)
-static const s_real rho_c = 322;     // Critical density (kg/m^3)
-static const s_real P_c = 2.2064e4;  // Critical Pressure (kPa)
+const s_real T_c = 647.096;   // Critical T (K)
+const s_real rho_c = 322;     // Critical density (kg/m^3)
+const s_real P_c = 2.2064e4;  // Critical Pressure (kPa)
 
 // To generalize the equation of state there are parmeters to set the number
```
Comment (blnicho) [inline src/helmholtz/iapws95/iapws95_param.h]: Typo: parameters

File: src/helmholtz/swco2/swco2_guess.cpp [hunk 34]
```diff
@@ -0,0 +1,176 @@
...
+    0.62385555*pow(tt, 0.5) -
+    0.32731127*pow(tt, 10.0/6.0) +
+    0.39245142*pow(tt, 11.0/6.0)
+  );
+}
+
+s_real delta_p_tau_supercritical_guess_swco2(s_real p, s_real tau){
+
+  /* This uses approximate isochors to get a reasonable initial guess for the
+  density solvution for a vapor at a particular temperature and pressure.  If
```
Comment (blnicho) [inline src/helmholtz/swco2/swco2_guess.cpp]: Typo: solution

File: src/helmholtz/swco2/swco2_guess.cpp [hunk 35]
```diff
@@ -0,0 +1,176 @@
...
+  else if(p < -202040.1487 + 782.056835*T - 0.144729326*T*T){ //rho < 900
+    return delta_p_tau_rf(p, tau, 795, 905, 0);
+  }
+  return 1180/rho_c;
+}
+
+s_real delta_p_tau_vap_guess_swco2(s_real p, s_real tau){
+
+  /* This uses approximate isochors to get a reasonable initial guess for the
+  density solvution for a vapor at a particular temperature and pressure.  If
```
Comment (blnicho) [inline src/helmholtz/swco2/swco2_guess.cpp]: Typo: solution

File: src/helmholtz/swco2/swco2_guess.cpp [hunk 36]
```diff
@@ -0,0 +1,176 @@
...
+  }
+  else if(p < -35876.37569 + 146.8404454*T - 0.012537284*T*T){ //rho < 400
+    return delta_p_tau_rf(p, tau, 295, 405, 0);
+  }
+  return delta_p_tau_rf(p, tau, 395, rho_c, 0);
+}
+
+s_real delta_p_tau_liq_guess_swco2(s_real p, s_real tau){
+  /* This uses approximate isochors to get a reasonable initial guess for the
+  density solvution for a vapor at a particular temperature and pressure.  If
```
Comment (blnicho) [inline src/helmholtz/swco2/swco2_guess.cpp]: Typo: solution

File: src/helmholtz/swco2/swco2_param.h [hunk 37]
```diff
@@ -0,0 +1,263 @@
...
+#define DELTA_VAP_SAT_GUESS delta_sat_v_approx(tau)
+
+const s_real R = 0.1889241;  // Specific gas constant (kJ/kg/K)
+
+// Critiacal point for water and R
+const s_real T_c = 304.128;   // Critical T (K)
+const s_real rho_c = 467.6;   // Critical density (kg/m^3)
+const s_real P_c = 7377.3;    // Critical Pressure (kPa)
+
+// To generalize the equation of state there are parmeters to set the number
```
Comment (blnicho) [inline src/helmholtz/swco2/swco2_param.h]: Typo: parameters

File: src/helmholtz/swco2/swco2_param.h [hunk 38]
```diff
@@ -0,0 +1,263 @@
...
+// of terms in each sumation.  So far we are looking at IAPWS95 and Span-Wagner
```
Comment (blnicho) [inline src/helmholtz/swco2/swco2_param.h]: Typo: summation

File: src/helmholtz/swco2/swco2_param.h [hunk 39]
```diff
@@ -0,0 +1,263 @@
...
+// They have the same types of terms but different numbers of term.  For both,
+// we are currently droping the nonanalytic terms, which we've labed S4 there
```
Comment (blnicho) [inline src/helmholtz/swco2/swco2_param.h]: Typo: dropping

File: src/helmholtz/common/helmholtz_external.cpp [hunk 40]
```diff
@@ -0,0 +1,444 @@
...
+        R*XD*XE_dd/XE/XE + 2.0*R*XD*XE_d*XE_d/XE/XE/XE;
+      hes[1] = cv_hes[1] + R*XD_dt/XE - R*XD_d*XE_t/XE/XE - R*XD_t*XE_d/XE/XE -
+        R*XD*XE_dt/XE/XE + 2.0*R*XD*XE_d*XE_t/XE/XE/XE;
+      hes[2] = cv_hes[2] + R*XD_tt/XE - 2.0*R*XD_t*XE_t/XE/XE -
+        R*XD*XE_tt/XE/XE + 2.0*R*XD*XE_t*XE_t/XE/XE/XE;}}
+  return cp(delta, tau);
+}
+
+s_real w_with_derivs(s_real delta, s_real tau, s_real *grad, s_real *hes){
+  //TODO<jce> hey you forgot one
```
[the remark this answers is not in the document]
Comment (eslickj) [reply src/helmholtz/common/helmholtz_external.cpp]: Yeah, it's still todo.  I should really get that done before someone tried to use the speed of sound somewhere.

File: src/helmholtz/common/helmholtz_config.h [hunk 41]
```diff
@@ -71,5 +74,18 @@
   #define s_fabs fabsl
 #endif
 
+inline void zero_derivs2(s_real *grad, s_real *hes){
+  // Set the derivatives to zero
+  if(grad!=NULL){
```
Comment (mengleCMU) [inline src/helmholtz/common/helmholtz_config.h]: Error during compiling = 'NULL' was not declared in this scope

Comment (mengleCMU) [reply src/helmholtz/common/helmholtz_config.h]: This isn't compiling not quite sure why, there are NULLs other places in the code

Comment (eslickj) [reply src/helmholtz/common/helmholtz_config.h]: @mengleCMU, give it a shot now.  I didn't test it but I did include stdlib, which should include the definition of NULL.  I restructured a lot of stuff, and it looks like depending on your system things may behave a bit different.

File: src/helmholtz/common/helmholtz_external.cpp (hunk 40 above)
[the remark this answers is not in the document]
Comment (eslickj) [reply src/helmholtz/common/helmholtz_external.cpp]: Okay, after more than a year, I finally finished the speed of sound derivatives.

**Comment by eslickj** [comment]
Okay, all the comments are addressed.  @mengleCMU let me know if you still have trouble compiling it.

@andrewlee94 and @ksbeattie.  How do you guys feel about doing a new binary beta release, but not updating the default location in IDAES just yet?  I want to build this once it's merged so people don't have too, but I want to do a lot more testing before we officially move to the new libraries.

**Comment by andrewlee94** [comment]
@eslickj A beta build of the binaries sounds like a good idea.

File: src/helmholtz/common/helmholtz_config.h (hunk 41 above)
Comment (mengleCMU) [reply src/helmholtz/common/helmholtz_config.h]: This now compiles

**Pull request merged**